flexmeasures/data/migrations/versions/f0ee99278f6f_added_flexmodel_to_generic_asset_model.py

```diff
@@ -51,57 +125,167 @@ def upgrade():
-        if len(sensors_with_key) > 1:
-            raise Exception(
-                f"Multiple sensors with 'soc-min' found for asset_id {asset_id}: {[s.id for s in sensors_with_key]}"
-            )
+    for field_spec in fields_specs:
+        old_name = field_spec["old_field_name"]
+        new_name = field_spec["new_field_name"]
+
+        # Check if the key exist on any asset's attributes and move that into the asset's flex_model
+        # This is to ensure that the flex_model is not empty in the edge case where the asset has no
+        # sensors or has no sensors using the field/key
+        asset_result = conn.execute(
+            sa.select(
+                generic_asset_table.c.id,
+                generic_asset_table.c.attributes,
+                generic_asset_table.c.flex_model,
+            ).where(generic_asset_table.c.attributes[old_name].isnot(None))
+        )
+        affected_assets = asset_result.fetchall()
+
+        for asset in affected_assets:
+            asset_id = asset.id
+            asset_attrs = asset.attributes or {}
+            flex_model_data = asset.flex_model or {}
 
-        sensor = sensors_with_key[0]
-        if sensor.attributes.get("soc-min") is not None:
             # check if value is a int or float
-            if not isinstance(sensor.attributes.get("soc-min"), (int, float)):
+            if not isinstance(asset_attrs[old_name], (int, float)):
                 raise Exception(
-                    f"Invalid value for 'soc-min' in sensor {sensor.id}: {sensor.attributes['soc-min']}"
+                    f"Invalid value for '{old_name}' in generic asset {asset_id}: {asset_attrs[old_name]}"
                 )
-            soc_min_value_kwh = sensor.attributes.get("soc-min") * 1000
+
+            soc_min_value_kwh = asset_attrs[old_name] * 1000
             soc_min_in_kwh = f"{soc_min_value_kwh} kWh"
-            flex_model_data = {"soc-min": soc_min_in_kwh}
+            flex_model_data[new_name] = soc_min_in_kwh
 
+            # Update the generic asset attributes to remove 'old_name' and add 'new_name' to flex_model
+            asset_attrs.pop(old_name, None)
             stmt = (
                 generic_asset_table.update()
                 .where(generic_asset_table.c.id == asset_id)
-                .values(flex_model=flex_model_data)
+                .values(
+                    flex_model=flex_model_data,
+                    attributes=asset_attrs,
+                )
             )
-
             conn.execute(stmt)
 
-            # Update the sensor attributes to remove 'soc-min'
-            sensor.attributes.pop("soc-min", None)
-            stmt = (
-                sensor_table.update()
-                .where(sensor_table.c.id == sensor.id)
-                .values(attributes=sensor.attributes)
-            )
-            conn.execute(stmt)
+        # Process the grouped sensors
+        for asset_id, sensors_with_key in field_spec["grouped"].items():
```

Comment:
With my earlier suggestion to use `field_spec["grouped"][sensor_generic_asset].append(sensor)` this could become:

```suggestion
        for sensor_generic_asset, sensors_with_key in field_spec["grouped"].items():
```
and you can avoid having to fetch the asset again.